network: add rtcp parser

chemag opened the pull request.

chemag commented on network/rtcp_packet.ksy:
```diff
@@ -0,0 +1,287 @@
+        repeat: expr
+        repeat-expr: sc_count
+    instances:
+      sc_count:
```
The RFC:

>    source count (SC): 5 bits
>       The number of SSRC/CSRC chunks contained in this SDES packet.  A
>       value of zero is valid but useless.


Did 's/sc_count/source_count/g'

GreyCat replied:
Well, the general guidance in the style guide is to keep these things consistent across all .ksy files in our repo to make it easier to understand for people without reading the official specs. If you want to convey original IDs, typically that's what `-orig-id` is for — it is totally ignored by the compiler, but you can it can appear in the generated documentation and it would be still available for search in the spec file. Moreover, there are no rules at all checked for `-orig-id` — it can contain spaces, lower/upper case chars, etc. You can do:

```yaml
seq:
  - id: num_source_chunks
    -orig-id: source count (SC)
    type: b5
```

chemag replied:
Note that `source_count` is a derived field. 
```
  sdes_packet:
    seq:
      - id: source_chunk
        type: source_chunk
        repeat: expr
        repeat-expr: source_count
    instances:
      source_count:
        value: _parent.subtype
```

Adding `orig-id` to an `instance` field craps out: 

```
$ ksc -t python rtcp_packet.ksy
/types/sdes_packet/instances/num_source_chunks/orig-id: invalid key found in value instance, allowed: doc, doc-ref, enum, if, value
Makefile:5: recipe for target 'rtcp_packet.py' failed
make: *** [rtcp_packet.py] Error 2
```

GreyCat replied:
It's `-orig-id`, not `orig-id`.

chemag commented on network/rtcp_packet.ksy:
```diff
@@ -0,0 +1,287 @@
+        value: b2 & 0x1fff
+      symbol_list:
+        if: t == 1
+        value: b2 & 0x3fff
```
Can you please discuss what you mean with this?

chemag replied:
I'd need to get b13 and b14 for this. I checked ip fragmentation on ipv4 (which should be very similar), but it's not implemented.

GreyCat replied:
b13 and b14 should work too, given that it's all big-endian stuff?

chemag replied:
b13 seems unsupported:

```
$ ksc -t python rtcp_packet.ksy
(main): /types/packet_status_chunk/instances/rle/value: unable to access 'b13' in rtcp_payload::packet_status_chunk context
```

GreyCat replied:
Um, this thing works for me:

```yaml
meta:
  id: foo
seq:
  - id: bar
    type: b13
```

Judging from the error message, you seem to be using `b13` as attribute name reference in `value` key, not as type.

chemag replied:
Forget about it. It definitely works

Done

chemag commented on network/rtcp_packet.ksy:
```diff
@@ -0,0 +1,287 @@
+      padding:
+        value: (b1 & 0x20) >> 5
+      subtype:
+        value: b1 & 0x1f
```
Great! Didn't realize you could do that. Done

chemag commented on network/rtcp_packet.ksy:
```diff
@@ -0,0 +1,287 @@
+    repeat: eos
+
+types:
+  rtcp_packet:
```
I agree the naming is unfortunate, but I need the different levels. Right now there is an UDP packet that contains an RTCP payload. This RTCP payload is composed of 1+ RTCP packets ("rtcp packet" is the term used in the RFC). What about using "rtcp_payload" for the outer abstraction, and "rtcp_packet" for the inner one?

GreyCat replied:
Ok, understood and totally agree then. Alternatively, you might want to consider "rtcp_packets" + "rtcp_packet", but I like your proposal better.

chemag replied:
done

chemag commented:
ptal

chemag commented:
ptal

chemag commented:
BTW, as a suggestion, `b1` is always interpreted as a Boolean. This means that, in a `.ksy` file, you have to access a `b1` field using `true` or `false`. This is not necessarily what the protocol expects: In many cases, the field is interpreted (per the standard) as 1 or 0. Is it possible to fix this?

This is not the

GreyCat commented:
> BTW, as a suggestion, `b1` is always interpreted as a Boolean. This means that, in a .ksy file, you have to  `b1` field using true or false. This is not necessarily what the protocol expects: In many cases, the field is interpreted (per the standard) as 1 or 0. Is it possible to fix this?

It boils down to how you would want to use it. Given that `b1` has only 2 possible values, and vast majority of these things are treated as flags, it's only natural to treat them as booleans in the first place, no avoid people always doing `if: flag != 0` instead of just `if: flag`.

If you really want to use them as 0 and 1, you can always convert them to integer by `.to_i`, i.e. something like:

```yaml
instances:
  foo:
    value: 123 + flag.to_i * 345
```

In this example, we want to use that as a number in some integer arithmetics, so we use `flag.to_i`, not just `flag`. But, obviously, given only 2 values, you can always write it like that as well:

```yaml
instances:
  foo:
    value: '123 + (flag ? 345 : 0)'
```

GreyCat commented:
@chemag Um, this PR seems to be messed up. Right now I see that you've (accidentally?) merged other PRs commit https://github.com/kaitai-io/kaitai_struct_formats/pull/112/commits/1afa4c25cdcec454d2ca59b123b6d4913c1fa429, and this commit https://github.com/kaitai-io/kaitai_struct_formats/pull/112/commits/620e6439dd7828b146a98d8845c960e1d56388ff actually adds `rtcp_payload.ksy` without removing `rtcp_packet.ksy`.

I can just extract `rtcp_payload.ksy` from this branch and apply it manually, or you can fix this branch (remember that you can use force push to overwrite everything) and I can merge the PR then. Which would you prefer?

chemag commented:
> Um, this PR seems to be messed up. Right now I see that you've (accidentally?) merged other PRs commit 1afa4c2, and this commit 620e643 actually adds rtcp_payload.ksy without removing rtcp_packet.ksy.

Hopefully I fixed it now.

KOLANICH commented:
>Add new module uImage

I'm not sure it belongs to this PR.

chemag commented:
> I'm not sure it belongs to this PR.

Hmm... I thought I had fixed this: 

```
$ git lall -20
* 3820f12 (HEAD -> master, origin/master, origin/HEAD) network: add rtcp parser
* 0a65250 Add new module uImage
* 50fa4ca Added security category + fixed title case
* 83bf172 openpgp_message: Fix YAML
* 9b199c1 Style adjustments
* 0831b6f gettext_mo: added doc
* c69b733 Moved gettext databases to databases, where it belongs to
*   2e249cb Merge pull request #96 from KOLANICH/gettext_mo
|\  
| * fe829c6 Added gettext mo
* | e5c7ade dex: Use preformatted docs everywhere
* | ea92987 dex: added title + initial doc
|/  
* 680a453 Renamed file to match ID
*   e442dd8 Merge pull request #104 from kolen/add-trd
|\  
| * 15049fb Add text about .trd files can be smaller than actual disk
| * d2b6386 Add file name indirection to extract first byte
| * 626ac32 WIP make all strings byte arrays
| * 51f2116 fixup! Add .trd (TR-DOS disk image) format parser
| * b3ab898 fixup! Add .trd (TR-DOS disk image) format parser
| * 04db5c8 Add .trd (TR-DOS disk image) format parser
* | 0f3febe Slight grammar fix

$ git push --force origin HEAD
Everything up-to-date
```

chemag commented:
OK, tried again. Hopefully this time is right :)

GreyCat commented:
Looks legit, thanks! Merging in, everything else could be addressed as extra PRs / commits.

chemag commented:
Thanks a lot for your help, guys!
